src/main/java/fi/metatavu/metamind/server/rest/StoriesApiImpl.java

```diff
@@ -266,45 +322,78 @@ public Response createSession(fi.metatavu.metamind.rest.model.Session body, UUID
       return createBadRequest("Invalid story parameter");
     }
     
-    UUID loggedUserId = getLoggerUserId();
     String locale = body.getLocale();
     String visitor = body.getVisitor();
     
-    // TODO: Permission check
-    
     Session session = sessionController.create(story, locale, body.getTimeZone(), visitor, loggedUserId);
     
+    try {
+      UUID createdResourceId = authenticationController.createProtectedResource(loggedUserId, String.format("Session-%s", session.getId()), String.format("/v2/stories/sessions/%s", session.getId()), "session", SCOPES);
+      List<UUID> policyIds = authenticationController.updatePermissionUsers(REALM_NAME);
+      authenticationController.upsertScopePermission(REALM_NAME, createdResourceId, SCOPES, String.format("Permission for session-%s", session.getId()), DecisionStrategy.AFFIRMATIVE, policyIds);
+      Set<UUID> permittedUsersId = authenticationController.getResourcePermittedUsers(REALM_NAME, createdResourceId, String.format("Session-%s", session.getId()), SCOPES);
+      if (!permittedUsersId.contains(loggedUserId) || loggedUserId == null) {
+        return createBadRequest(String.format(NO_USER_FOUND_MESSAGE));
+      }
+    } catch (Exception e) {
+      logger.warn(String.format(EXCEPTION_CAUGHT, e.getMessage()));
+    }
+    
     return createOk(sessionTranslator.translateSession(session));
   }
 
   @Override
   public Response createStory(Story body) {
-    UUID loggedUserId = getLoggerUserId();
+
     Locale locale = LocaleUtils.toLocale(body.getLocale());
     
-    // TODO: Permission check
-
-    return createOk(storyTranslator.translateStory(storyController.createStory(locale, body.getName(), body.getDafaultHint(), loggedUserId)));
+    fi.metatavu.metamind.persistence.models.Story story = storyController.createStory(locale, body.getName(), body.getDafaultHint(), loggedUserId);
+    try {
+      UUID createdResourceId = authenticationController.createProtectedResource(loggedUserId, String.format(STORY_NAME_BEGINNING, story.getId()), String.format("/v2/stories/%s", story.getId()), "story", SCOPES);
+      List<UUID> policyIds = authenticationController.updatePermissionUsers(REALM_NAME);
+      authenticationController.upsertScopePermission(REALM_NAME, createdResourceId, SCOPES, String.format("Permission for story-%s", story.getId()), DecisionStrategy.AFFIRMATIVE, policyIds);
+      Set<UUID> permittedUsersId = authenticationController.getResourcePermittedUsers(REALM_NAME, createdResourceId, String.format(STORY_NAME_BEGINNING, story.getId()), SCOPES);
+      if (!permittedUsersId.contains(loggedUserId) || loggedUserId == null) {
+        return createBadRequest(String.format(NO_USER_FOUND_MESSAGE));
+      }
+    } catch (Exception e) {
+      logger.warn(String.format(EXCEPTION_CAUGHT, e.getMessage()));
```

Comment:
When unhandled exception is encountered, internal error (500) should be returned, not logged as a warning